### PR TITLE
Per-path authentication

### LINKED ISSUE #82: Per-path HTTP authentication
Adding a mechanism wherein any subpath could be (un)protected independently of the global config.

### PR DESCRIPTION
**Pull request by nabijaczleweli**
1. Add `--{gen-,}path-auth` on top of `--{gen-,}auth`, which allows arbitrary nesting of basic auth credentials, and deprecate `--{gen-,}auth`. Closes #82
2. Dep bump and global `try!()` -> `?` modernisation
3. Remove needless allocations when handling paths.

<sub>Note: do not merge this with a merge commit. Do an FF merge if you need to but better yet leave this to me and I'll do it cleanly.</sub>

### CODE REVIEW

**Comment by nabijaczleweli** [comment]
If you've got a good take on how to improve the opening status line(s) to reduce cramping, I'm all for't

File: http.md [hunk 1]
```diff
@@ -227,7 +252,8 @@ pass parameters like what port to use.
 
   `http --auth Pirate:memelord42`
 
-    As in the first example, but require all clients to log in with the username "Pirate" and password "memelord42".
```
Comment (thecoshman) [inline http.md]: Why we splitting these lines, and why at such random point? You're not splitting other long lines.

**Comment by thecoshman** [comment]
the 'global' auth is basically just path-auth, but with the path set to `/`. In which case, I think we can avoid having `path-auth` and `gen-path-auth`, instead, just make the `<path>=` part optional, defaulting to `/`. 

I believe that would be none-breaking. Of course, we'd still have the error handling so that only a single authentication parameter can for a specific path.

Also, have you verified you can do things like have an authenticated folder, with an public sub-folder with an authenticated sub-folder? something like:

```
/foo # no authentication
/foo/bar # authenticated
/foo/bar/win # not authentication
```

**Comment by thecoshman** [comment]
> 
> 
> If you've got a good take on how to improve the opening status line(s) to reduce cramping, I'm all for't

I'm not sure what you mean... what's this in reference to?

File: http.md (hunk 1 above)
Comment (nabijaczleweli) [reply http.md]: Fixed

**Comment by nabijaczleweli** [comment]
1. Hmm if you add the path segment to `--auth`, that's a breaking change, because, e.g. people's credentials might've contained an equal sign. So, if you're down to make a breaking release, I could merge down the path interface into `--auth` so you can set the root password via `--auth /=UNAME:PWD`, or if you aren't, I could allow the root password to be set via `--path-auth` and squish it into `--auth` on a later breaking release (the former is preferable). As for explicitly having a root password in-code, I'd rather not remove it because (a) the root password is most likely to be set when the others aren't, and it's much faster to not look through the subpaths, and (b) the loop control gets iffier.

2. Yes that path/auth scheme works.

3. What I'm referring to is that you get big blurb with many words like this sometimes:

> Hosting "." on port 8000 without TLS and basic global authentication using "admin" as username and "admin" as password, no authentication under "target/.rustc_info.json", basic authentication under "target/debug" using "r2ciYq00k8a" as username and "Z-zn~HV,R`<{br/0Yf." as password, no authentication under "target/doc" basic authentication under "target/release" using "releases" as username and no password...

And that you might have a better way of outputting it.

**Comment by thecoshman** [comment]
>     1. Hmm if you add the path segment to `--auth`, that's a breaking change, because, e.g. people's credentials might've contained an equal sign. So, if you're down to make a breaking release, I could merge down the path interface into `--auth` so you can set the root password via `--auth /=UNAME:PWD`, or if you aren't, I could allow the root password to be set via `--path-auth` and squish it into `--auth` on a later breaking release (the former is preferable). As for explicitly having a root password in-code, I'd rather not remove it because (a) the root password is most likely to be set when the others aren't, and it's much faster to not look through the subpaths, and (b) the loop control gets iffier.

Hmm... didn't think about the parsing hell with people having `=` in their passwords. I don't think it's worthwhile making this a breaking change. If we have some other change that really needs to be made and should be made breaking, then sure, we can redo this interface to make it easier to use.

>     2. Yes that path/auth scheme works.

nice!

>     3. What I'm referring to is that you get big blurb with many words like this sometimes:
> 
> 
> > Hosting "." on port 8000 without TLS and basic global authentication using "admin" as username and "admin" as password, no authentication under "target/.rustc_info.json", basic authentication under "target/debug" using "r2ciYq00k8a" as username and "Z-zn~HV,R`<{br/0Yf." as password, no authentication under "target/doc" basic authentication under "target/release" using "releases" as username and no password...
> 
> And that you might have a better way of outputting it.

Those are a series of strings that describe the permissions that have been selected; simply put each one on a new line, no?

**Pull request merged**